src/typing/constraint_js.ml

```diff
@@ -529,21 +529,34 @@ let copy_node node = match node with
     Root { rank; constraints = Unresolved (copy_bounds bounds) }
   | _ -> node
 
+type scope_kind =
+  | VarScope (* var, functions hoisted up to this point *)
+  | LexicalScope (* let, const, classes *)
+
 type scope_entry = {
   specific: Type.t;
   general: Type.t;
   def_loc: Spider_monkey_ast.Loc.t option;
   for_type: bool;
+  constant: bool;
+  scope_kind: scope_kind;
```

Comment:
It turns out the behavior of let-declarations in var-scope I showed above is from an older version of the spec. Newer versions have [simpler semantics](https://people.mozilla.org/~jorendorff/es6-draft.html#sec-function-definitions-static-semantics-early-errors)

    It is a Syntax Error if the LexicallyDeclaredNames of FunctionStatementList contains any duplicate entries.

Doesn't change the fact that I should clarify the existing comments, add more comments, and work on Jeff's suggestion!